### ISSUE TITLE
Friendlier search macros?

### ISSUE THREAD

**Issue by ferd** [proper-testing/proper#157]
Guided by Kostis' keynote this morning at BEAMCode SF, I was reading from the source and some docs, and noticed that:

a) the first method suggested in the talk:

```erlang
?FORALL_SA(GVar, #{gen => MyGen},
    begin
        some_calc(GVar),
        ?MAXIMIZE(GVar),
        my_prop(GVar)
    end).
```
is currently deprecated according to code comments.

b) The alternative format mentioned is instead:
```erlang
?EXISTS(GVar, #{gen => MyGen},
    begin
        some_calc(GVar),
        ?MAXIMIZE(GVar),
        my_prop(GVar)
    end).
```
With a similar variation existing for `?NOT_EXISTS`, but suddenly nothing seems to be there for `?FORALL` that is not deprecated.

I'm curious what the intended usage long term should be. Is there a reason for abandoning the ?FORALL approach and aiming towards `EXISTS` or `NOT_EXISTS`?

 For me, and in a way of making the friendliest thing possible, I was thinking of a macro a bit like:

```erlang
?TARGET_ALL(GVar, RegularGenerator, % no #{gen => G} deal here
    begin
        ?MAXIMIZE(whatever(GVar)),
        my_prop(GVar)
    end).
```
I can imagine easily getting something similar to that with `NOT_EXISTS(GVar, #{gen => G}, not Prop)` but I'm a bit puzzled by the rationale behind the current approach.

**Comment by TheGeorge** [comment]
Hi,
thank you for your input. I can give some insight into why the macros are as they are:

a) TPBT was originally developed as an extension project to PropEr, something that you could plugin when you needed it but also something that wouldn't change how core PropEr is implemented. Like an Addon library. `?FORALL_SA` is actually a short form for:
```
?STRATEGY(simulated_annealing,
          ?FORALL(...)).
```
The `?TARGET` macro hands the generator (and possibly a neighborhood function) to the library, which performs some calculations, and returns a PropEr generator. This generator is then, of course, hooked up to the search strategy.

However, as you mentioned, this is a bit ugly and cumbersome. You need to know how the system works and the macros are not self-explanatory. 

b) The `?EXISTS` and `?NOT_EXISTS` macros where chosen because they allow the same expressiveness as the `?FORALL` macro and they are visually different from them:
 + `?NOT_EXISTS` is the targeted version of `?FORALL`, one just needs to invert the condition of the property.
 + `?EXISTS` is the targeted version of `proper:fails(?FORALL(...))`

So by reading the property, it should be which parts are randomly generated and which parts are explored by a search. So if we write something like this:
```
prop_maze() ->
  ?FORALL(Maze, maze_generator(),
         ?EXISTS(Path, #{gen => path(Maze)},
                  begin
                    ?MINIMIZE(distance_to_exit(Maze, Path)),
                    path_reaches_exit(Maze, Path)
                  end))
```
we can see that mazes are generated randomly, but the paths are generated by using a search strategy.

Now both of these macros still have the `#{gen => Generator}` thingy which I agree should be reduced to just `Generator`. Along with updating the documentation (so that it can be put online) that's one thing on our todo list.

TL;DR: The `?FORALL` stays and is used for random property-based testing and the `?(NOT_)EXISTS` is the new macro for targeted property-based testing and the two are combinable.

By the way, there is a tutorial on the [PropEr homepage](http://proper.softlab.ntua.gr/Tutorials/PropEr_testing_with_search_strategies.html).

**Comment by ferd** [comment]
Thanks for the clarifications! So one thing that was not obvious to me from the code was whether `?EXISTS` and `?NOT_EXISTS` should be nested within `?FORALL` at all times; they appear to work fine outside of it (albeit things like iteration limits seem to be ignored in that context). So I wanted to be sure to cover that properly.

When I am nesting a search within a `?FORALL`, is the search iterating multiple times per `?FORALL` iteration? If I used `?EXISTS` alone, it can do many iterations as it pleases. So if I'm nesting it, and I specify 100 runs, am I getting `100*SearchCount` total executions or does the constraint propagate down and I get 100 flat?

**Comment by TheGeorge** [comment]
They work fine outside but have their own counter which can be set in the options. There are two new options for proper: 
 + `search_steps` and 
 + `search_strategy` to specify which search strategy you want (currently there is only sa, but that might change)
(Again, sorry for not keeping up with the documentation)

When you nest an `?EXIST` inside a `?FORALL`  then `?FORALL` creates `numtests` tests and for each of those tests the `?EXISTS` can run up to `search_steps` tests. The reason here is that you want to have control over how thorough the search is. 

If we would use a global number, later tests might have not enough tests left to perform a good search. If we would use the same number twice (`numtests*numtests`) we are faced with a situation where we can run either a lot of tests with a thorough search or a few tests with a shallow one, but one might run a few tests with a thorough search or vice versa.

**Comment by ferd** [comment]
Alright, thanks.

I'll try to update the rebar3 plugin for PropEr to handle these options. Any idea when these changes might make it to a Hex release?

**Comment by TheGeorge** [comment]
I think this can be closed (see #158 )

**Comment by kostis** [comment]
Closing it.
